knebind/solver/solver.go

```diff
@@ -282,7 +338,7 @@ func (s *solver) solve() (*assign, error) {
 				}
 			}
 		}
-		port2IntfChan := genCombos(port2Intfs)
+		port2IntfChan := genCombos(port2Intfs, s.checkPort)
```

Comment:
I'm also wondering if a key problem here is that we're searching through port combinations rather than _link_ combinations. If instead of creating port2Intfs we could create a link2Links maps that takes device, port, and link into consideration. Then there would need to be one smaller added phase to map the unlinked ports. Would that give you at least the same effective pruning that this accomplishes and perhaps more?

Reply:
Yes, the key issue is the port2Intfs (link2Links makes it better). However, I think, the generic issue is generating all solution sets some of which may be invalid. So whether for DUT or PORT if at any iteration, the key -> value selection may lead to invalid solutions lets not proceed down that branch. Also, I think, the PORT selection should be in conjunction with DUT selection (not two separate process); because that will prune a lot of invalid combinations for larger topologies.